fx quant: fix fq when input is quantized and node does not need fq

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49420 fx quant: make sure observer is inserted before a quantized output
* **#49382 fx quant: fix fq when input is quantized and node does not need fq**
* #49239 fx quant: do not insert observers at quantized inputs
* #49238 fx quant: move {input|output}_quantized_idxs cfg from convert to prepare
* #49213 eager quant: remove fake_quant after add/mul nodes during QAT

Summary:

Fixes an edge case.  If the input to the graph is quantized and the
first node does not need activation observation, makes sure that
the observer is not inserted.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_int8_input_no_unnecessary_fq
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25551041](https://our.internmc.facebook.com/intern/diff/D25551041)

vkuzo committed:
fx quant: fix fq when input is quantized and node does not need fq

Summary:

Fixes an edge case.  If the input to the graph is quantized and the
first node does not need activation observation, makes sure that
the observer is not inserted.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_int8_input_no_unnecessary_fq
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -438,6 +438,11 @@ def load_arg(a):
+                    # elsewhere. For now, make it as observed for the purposes
+                    # of this file.  If in the future we will need the observer
+                    # instance for anything, we will have to adjust the API
+                    observed_node_names_set.add(node.name)
```
oh, should this be merged with previous pR

jerryzh168 replied:
I think we should always do this. What are the cases that needs observer that might come up in the future?

vkuzo replied:
> I think we should always do this. What are the cases that needs observer that might come up in the future?

np, removed the comment

> oh, should this be merged with previous pR

I don't feel too strongly, but technically this is a separate bug and fix.  The previous PR removes the observer on the input of the first node, if the input is quantized.  This PR removes the observer on the output of the first node, if the input is quantized and the node does not need observation.  Lmk if you prefer them to be combined.

jerryzh168 replied:
sg, I don't have a strong preference either.

vkuzo committed:
Update on "fx quant: fix fq when input is quantized and node does not need fq"


Summary:

Fixes an edge case.  If the input to the graph is quantized and the
first node does not need activation observation, makes sure that
the observer is not inserted.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_int8_input_no_unnecessary_fq
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25551041](https://our.internmc.facebook.com/intern/diff/D25551041)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1524,6 +1524,32 @@ def forward(self, x):
+                return self.add_func.add_scalar(x, self.scalar)
+
+        m = M(0.5)
+        d = {}
```
remove?

vkuzo committed:
comments on "fx quant: fix fq when input is quantized and node does not need fq"


Summary:

Fixes an edge case.  If the input to the graph is quantized and the
first node does not need activation observation, makes sure that
the observer is not inserted.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_int8_input_no_unnecessary_fq
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25551041](https://our.internmc.facebook.com/intern/diff/D25551041)

[ghstack-poisoned]

vkuzo committed:
rebase on "fx quant: fix fq when input is quantized and node does not need fq"


Summary:

Fixes an edge case.  If the input to the graph is quantized and the
first node does not need activation observation, makes sure that
the observer is not inserted.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps.test_int8_input_no_unnecessary_fq
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25551041](https://our.internmc.facebook.com/intern/diff/D25551041)

[ghstack-poisoned]
